control-plane-operator/controllers/hostedcontrolplane/cco/reconcile.go

```diff
@@ -0,0 +1,454 @@
+package cco
+
+import (
+	"bytes"
+	"path"
+	"text/template"
+
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/pki"
+	prometheusoperatorv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/imageprovider"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/kas"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	"github.com/openshift/hypershift/support/certs"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/metrics"
+	"github.com/openshift/hypershift/support/proxy"
+	"github.com/openshift/hypershift/support/util"
+)
+
+const (
+	workerNamespace                = "openshift-cloud-credential-operator"
+	workerServiceAccount           = "cloud-credential-operator"
+	kubeconfigFile                 = "token"
+	metricsPort                    = 8443
+	operatorStartScriptTemplateStr = `#!/bin/bash
+set -euo pipefail
+
+while true; do
+   if [[ -f {{ .TokenDir }}/{{ .TokenFile }} ]]; then
+      break
+   fi
+   echo "Waiting for service account token"
+   sleep 2
+done
+
+echo token:
+cat {{ .TokenDir }}/{{ .TokenFile }}
+echo ===
+
+exec /usr/bin/cloud-credential-operator operator --kubeconfig {{ .TokenDir }}/{{ .TokenFile }}
+`
+	rbacProxyStartScriptTemplateStr = `#!/bin/bash
+set -euo pipefail
+
+while true; do
+   if [[ -f {{ .TokenDir }}/{{ .TokenFile }} ]]; then
+      break
+   fi
+   echo "Waiting for service account token"
+   sleep 2
+done
+
+exec /usr/bin/kube-rbac-proxy \
+    --secure-listen-address=0.0.0.0:{{ .MetricsPort }}" \
+    --upstream=http://127.0.0.1:2112/ \
+    --tls-cert-file={{ .CertFile }} \
+    --tls-private-key-file={{ .KeyFile }}
+`
+)
+
+var startScript string
+var proxyStartScript string
+
+func init() {
+	var err error
+	startScriptTemplate := template.Must(template.New("script").Parse(operatorStartScriptTemplateStr))
+	startScript, err = renderOperatorStartScript(startScriptTemplate)
+	proxyStartScriptTemplate := template.Must(template.New("script").Parse(rbacProxyStartScriptTemplateStr))
+	proxyStartScript, err = renderProxyStartScript(proxyStartScriptTemplate)
+	if err != nil {
+		panic(err.Error())
+	}
+}
+
+func renderTokenWait(templ *template.Template) (string, error) {
```

Comment:
where is this used?

Reply:
This is transient and deleted, please ignore.